### DISCUSSION TITLE
Does fcitx5 provide the feature of calling scripts when switching input methods?

### DISCUSSION THREAD

**Discussion by paradoxskin** [fcitx/fcitx5#919] [Q&A] [answered]
I've developed a Python script to manage dwm's status bar. To optimize CPU usage in status retrieval, I've set up a refresh rate of once per minute for time, battery, and Wi-Fi data. Most status updates are monitored via a watchdog system that tracks file changes. These files are altered solely by other scripts upon deliberate status changes.
![2023-12-21_01-38](https://github.com/fcitx/fcitx5/assets/45301314/a8922a3d-8edf-4b97-b0b1-a938906cd179)
To display fcitx5's real-time status, I've implemented a thread that polls fcitx5-remote every 0.2 seconds to retrieve the current input method status. 

```python
def get_im():
    global value, display
    while True:
        state = system("fcitx5-remote")
        if state != value["fcitx5"]:
            value["fcitx5"] = state
            display["fcitx5"] = [' ', ' ', '^b#fb8feb^^c#443326^󱌱^d^'][int(state)]
            gen()
        time.sleep(.2)
```
However, this approach seems inefficient, considering that changing the input method is essentially a user-initiated action. My initial strategy involves using file alterations or local sockets to communicate with fcitx5, similar to other status updates. Regrettably, I haven't found any online resources detailing methods for fcitx5 to execute specific scripts when switching input methods.

Does fcitx5 support a similar functionality natively? If not, how feasible would it be to modify its source code to incorporate this feature?

**Comment by wengxt** [comment, marked as answer]
You can use fcitx5-lua to monitor certain internal event

https://fcitx.github.io/fcitx5-lua/modules/fcitx.html#watchEvent
for your use case, you can focus on FocusIn and SwitchInputMethod

https://fcitx.github.io/fcitx5-lua/modules/fcitx.html#currentInputMethod
